### PR TITLE
Quantile Q-Learning Implementation

### PR DESCRIPTION
**Pull request by frederikschubert**
This PR adds a `QuantileQ` class with function types 3 and 4 that accept a number of quantiles together with the state (and action), as well as a `QuantileQLearning` class. The `QuantileQ` function could be merged into the `Q` class which would simplify the user-facing API. However, some more work needs to be done to incorporate the `QuantileQLearning` class into the `QLearning` class. I just wanted to validate that this is the correct approach to take to implement the IQN.

There still is some documentation for the quantile huber loss missing and the notebooks need to be added and tuned.

Closes https://github.com/coax-dev/coax/issues/3

### CODE REVIEW

**Comment by frederikschubert** [comment]
@KristianHolsheimer I refactored everything to use the `StochasticQ` class by introducing an `EmpiricalQuantileDist` distribution that is parametrized by the samples from the quantile function and also carries the quantile fractions for the loss computation. The `StochasticQ` class uses quantile regression if no `value_range` is given. 

What do you think about this approach?

File: coax/_core/stochastic_q.py [hunk 1]
```diff
@@ -122,11 +130,12 @@ def num_bins(self):
 
     @classmethod
     def example_data(
-            cls, env, value_range, num_bins=51, observation_preprocessor=None,
-            action_preprocessor=None, value_transform=None, batch_size=1, random_seed=None):
+            cls, env, value_range, proba_dist_hparams=dict(num_bins=51, value_range=[-10, 10]),
```
Comment (KristianHolsheimer) [inline coax/_core/stochastic_q.py]: Please don't use mutable defaults. These could generate unexpected side-effects, e.g.
```python
def f(d={}):
    return d

x = f()  # Returns: {}
x['foo'] = 'bar'
f()  # Now returns: {'foo': 'bar'}
```
A typical pattern is to default to `None` and to set the default dict internally.

File: coax/__init__.py [hunk 2]
```diff
@@ -57,6 +57,7 @@
     # classes and functions
     'V',
     'Q',
+    'QuantileQ',
```
Comment (KristianHolsheimer) [inline coax/__init__.py]: Do we still need this after refactor?

File: coax/_base/test_case.py [hunk 3]
```diff
@@ -230,7 +230,7 @@ def func(S, is_training):
         return func
 
     @property
-    def func_q_tochastic_type1(self):
+    def func_q_stochastic_type1(self):
```
Comment (KristianHolsheimer) [inline coax/_base/test_case.py:233]: Thanks!

File: coax/_core/stochastic_q.py [hunk 4]
```diff
@@ -48,15 +48,22 @@ class StochasticQ(BaseStochasticFuncType1):
...
     num_bins : int, optional
 
-        The space of rewards is discretized in :code:`num_bins` equal sized bins. We use the default
-        setting of 51 as suggested in the `Distributional RL <https://arxiv.org/abs/1707.06887>`_
-        paper.
+        If `value_range` is given: The space of rewards is discretized in :code:`num_bins` equal
+        sized bins. We use the default setting of 51 as suggested in the
+        `Distributional RL <https://arxiv.org/abs/1707.06887>`_paper.
+
+        Else: The number of fractions of the quantile function of the rewards is defined by
```
Comment (KristianHolsheimer) [inline coax/_core/stochastic_q.py:63]: Thanks for expanding the docstring as well.

File: doc/examples/stubs/index.rst [hunk 5]
```diff
@@ -20,6 +20,7 @@ an agent.
     soft_qlearning
     dqn
     dqn_per
+    iqn
```
Comment (KristianHolsheimer) [inline doc/examples/stubs/index.rst:23]: Thanks for including the example!

File: coax/proba_dists/_empirical_quantile.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+from ..utils import jit, isscalar
+from ._base import BaseProbaDist
+
+
+__all__ = (
+    'EmpiricalQuantileDist',
+)
+
+
+class EmpiricalQuantileDist(BaseProbaDist):
```
Comment (KristianHolsheimer) [inline coax/proba_dists/_empirical_quantile.py:36]: Thanks for taking the time to refactor this to use StochasticQ!

**Review by KristianHolsheimer — CHANGES_REQUESTED** [review]
Thanks for taking the time to refactor this to use StochasticQ!

File: coax/__init__.py (hunk 2 above)
Comment (frederikschubert) [reply coax/__init__.py]: I missed the reference but removed it in a subsequent commit.

File: coax/_core/stochastic_q.py (hunk 1 above)
Comment (frederikschubert) [reply coax/_core/stochastic_q.py]: I also missed this in the initial refactoring commit but changed it afterwards.

File: coax/_core/stochastic_q.py (hunk 4 above)
Comment (frederikschubert) [reply coax/_core/stochastic_q.py:63]: I am glad I can help and get some experience with jax!

**Comment by frederikschubert** [comment]
Thank you for looking into my contribution!

I will still have to add a stub for the IQN and add some unit tests for the new functionality in `StochasticQ` and `QLearning`. Also I need to have a look at the documentation after the API is now finalized. But then this should be good to go.

Do you want me to squash the commits in the end so that the history in the main branch is clean?

**Comment by KristianHolsheimer** [comment]
> Thank you for looking into my contribution!
No worries and thank you for the well-written PR!

> I will still have to add a stub for the IQN and add some unit tests for the new functionality in StochasticQ and QLearning. Also I need to have a look at the documentation after the API is now finalized. But then this should be good to go.

Okay, shall we include that in this PR as well? Also, You probably noticed this yourself, but just in case... I noticed that the `doc/examples/cartpole/iqn.rst` script wasn't included in `doc/examples/cartpole/index.rst`. 

> Do you want me to squash the commits in the end so that the history in the main branch is clean?

Yes, will do.

**Review by KristianHolsheimer — APPROVED** [review]

**Comment by frederikschubert** [comment]
I added the stub, some documentation and a test that `StochasticQ` works with both quantile function types. Do you want me to add more tests or should we leave it as is? Otherwise I think this PR is ready to merge.

**Review by KristianHolsheimer — APPROVED** [review]
Thanks again for this well-written PR! Especially with the extra documentation.. highly appreciated!!

**Comment by frederikschubert** [comment]
No worries! I just fixed the quantile huber loss docs, because I saw that they were not included in the documentation and there was an error. So now it should be really good to go. :)

**Pull request merged**